### PR TITLE
feat: introduce constraint reduction

### PR DESCRIPTION
**Pull request by mlutze**

### CODE REVIEW

**Comment by magnus-madsen** [comment]
!!!jar

**Comment by magnus-madsen** [comment]
!!!jar

**Comment by mlutze** [comment]
!!!rerun

**Comment by mlutze** [comment]
!!!rerun

**Comment by mlutze** [comment]
@magnus-madsen Still some rough edges, but a review would help point out the most important ones to sand down.

**Comment by magnus-madsen** [comment]
The changes to unification and Typer cannot be done separately, I assume?

**Comment by magnus-madsen** [comment]
This effectively replaces the type system right? There is no flag

**Comment by magnus-madsen** [comment]
!!!jar

**Comment by magnus-madsen** [comment]
This program crashes which we need to deal with because it is common to type:

```
mod Foo {
    trait Add[a] {
        pub type Aef: Eff
        pub def add(x: a, y: a): a \ Add.Aef[a]
    }

    instance Foo.Add[t] {
        pub def add(x: t, y: t): t \ Aef[a] = ???
    }
}
```

**Comment by magnus-madsen** [comment]
This also crashes:

```
    instance Add[String] {
        type Aef = {}
        pub def add(x: String, y: String): String = x + y
    }
```

File: main/test/flix/experimental/Test.Dec.AssocEff.flix
Comment (magnus-madsen) [inline main/test/flix/experimental/Test.Dec.AssocEff.flix:1]: I will ask Jakob til help with some more tests.

File: main/test/flix/experimental/ExperimentalSuite.scala
Comment (magnus-madsen) [inline main/test/flix/experimental/ExperimentalSuite.scala:30]: Undo threads = 1

File: main/src/ca/uwaterloo/flix/language/phase/unification/Substitution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/Substitution.scala:192]: I would like this small change to be in a separate PR -- to gauge performance cost. Can you make one just for Substitution?

File: main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification.scala:127]: I would like this change to EffUnification to be in one small PR, to gauge performance cost.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification.scala:127]: And correctness under the old system.

File: main/src/ca/uwaterloo/flix/language/phase/Typer.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Typer.scala:20]: These changes effectively make the rest of the Typer dead code, correct?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Typer.scala:20]: We agree that the new system is used here / now, right? I see you continue with the found subst.

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:2]: 2024:)

**Comment by magnus-madsen** [comment]
I am doing a review. Do we agree that post-merge, the old typer can be removed, along with any cruft. E.g. old error messages, etc?

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:100]: Add some comment that now the stuff happens.

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:95]: maybe declaredTypConstraint and declaredEffConstraint?

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:102]: Where was the call to start logging?

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:106]: You say deferred, but I say pending or unresolved. I guess we should harmonize that, but probably not important now.

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:113]: Can we reorder the functions so the file can be read top to bottom?

Here is ok with the helper functions expandClassEnv and a few others, but then I think the main resolve function should come.

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:119]: linebreak

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:116]: Example of what this does? And is it done once?

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:147]: example?

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:428]: -> resultion result

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:428]: Reword sentence, it returns a Result which is ...

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:430]: I was wondering: the cev and eqEnv stay fixed right? So maybe we should make them implicit through all functions?

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:434]: prog-> progress

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:431]: Describe a bit what we are doing here.

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:450]: Explain that we actually break out of the whole thing.

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:539]: What are the components?

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:634]: I am OK with this for now, but whats the overall plan for this?

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:591]: Drop comment.

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:634]: Like whats the source of the messy nice? Is it that provenance is not enough? Or the constraint do not carry enough information? (which would be provenance, I guess)?

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:221]: Mismatched kinds?

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:226]: types -> proper types (or something with kind star)

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:241]: linebreak

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:244]: I know this is reflexivity, but it would be more natural to have as the first rule. imho. I leave it to you.

**Comment by magnus-madsen** [comment]
I still have some more code to review, but that is the first batch.

**Comment by magnus-madsen** [comment]
Is this not permitted:

```
    class A[a] {
        type Aef: Eff
        pub def f(x: a, y: a): a \ A.Aef[a]
    }

    class B[a] with A[a] {
        pub def flaf(x: a, y: a): a \ A.Aef[a]
    }
```

File: main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification.scala
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification.scala:127]: This breaks in the old system: https://github.com/flix/flix/pull/7275

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification.scala:127]: That's fucked up. But I see the real fault is in our region stuff.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification.scala:127]: Ok, alright. We leave it here.

**Comment by magnus-madsen** [comment]
Another kinding related bug. This one does not crash, but seems dangerous:

```
mod Foo {

    trait Add[t] {
        type Rhs: Type
        pub def add(lhs: t, rhs: Add.Rhs[t]): t
    }

    instance Add[Set[t]] with Order[t] {
        type Rhs = {Pure}
        pub def add(lhs: Set[t], rhs: t): Set[t] = Set.insert(rhs, lhs)
    
    }

}
```

**Comment by magnus-madsen** [comment]
In this monomorphic function:

```
    pub def main(): Unit \ IO = 
        let x = List#{1, 2, 3};
        let _y = ToJava.toJava(x);
        println("Hello World!")
```

the hover type of `_y` is `Output[List[Int32]]` which is correct, but unexpected. Why has it not been reduced to its concrete type?

Rest of program:

```
    trait ToJava[t] {
        type Output
        pub def toJava(t: t): ToJava.Output[t] \ IO
    }

    instance ToJava[List[t]] {
        type Output = ##java.util.List
        pub def toJava(t: List[t]): ##java.util.List \ IO = 
            Adaptor.toList(t)
    }

    pub def main(): Unit \ IO = 
        let x = List#{1, 2, 3};
        let _y = ToJava.toJava(x);
        println("Hello World!")
```

File: main/src/ca/uwaterloo/flix/language/phase/Typer.scala
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/Typer.scala:20]: Yes, everything in the old `TypeInference.scala` file dies.

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:102]: It was at the head of visitSpec. I'll add a compiler flag so we can specify which definition to log.

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:116]: It is done once per def/sig

**Comment by magnus-madsen** [comment]
One more (maybe an instance of the others, but I don't think so-- its not a kind error, its a type error):

```
trait Foo[t] {
    type K: Type -> Type
    type E: Type

    pub def f(x: t): Foo.K[t][Foo.E[t]]
}

instance Foo[Int32] {
    type K = List
    type E = String
    pub def f(x: Int32): List[String] = "${x}" :: Nil
}


instance Foo[Int64] {
    type K = Set
    type E = Char // OOPS
    pub def f(x: Int64): Set[String] = Set#{"${x}"}
}
```

**Comment by magnus-madsen** [comment]
An unexpected one:

```
def race(f: a -> b \ ef, l: Nel[a]): b \ ef = 
    region rc {
        let (tx, rx) = Channel.buffered(rc, Nel.length(l));
        foreach (x <- l) { 
            spawn (Channel.send(f(x), tx)) @ rc
        };
        Channel.recv(rx)
    }
```

**Comment by mlutze** [comment]
> This program crashes which we need to deal with because it is common to type:
> 
> ```
> mod Foo {
>     trait Add[a] {
>         pub type Aef: Eff
>         pub def add(x: a, y: a): a \ Add.Aef[a]
>     }
> 
>     instance Foo.Add[t] {
>         pub def add(x: t, y: t): t \ Aef[a] = ???
>     }
> }
> ```

The easiest way to address this is to make the MissingAssocDef error unrecoverable until we have a more robust way of dealing with ERROR types. Is that OK?

**Comment by magnus-madsen** [comment]
> > This program crashes which we need to deal with because it is common to type:
> > ```
> > mod Foo {
> >     trait Add[a] {
> >         pub type Aef: Eff
> >         pub def add(x: a, y: a): a \ Add.Aef[a]
> >     }
> > 
> >     instance Foo.Add[t] {
> >         pub def add(x: t, y: t): t \ Aef[a] = ???
> >     }
> > }
> > ```
> 
> The easiest way to address this is to make the MissingAssocDef error unrecoverable until we have a more robust way of dealing with ERROR types. Is that OK?

Sure

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:269]: Doesn't exist.

Added this to followup

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:683]: I don't think that provenance is good enough because there are many constraints that contribute to an error, but we consider whichever one we actually fail on to be the cause.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:683]: I see. But it seems hard to do better with a lot more stuff.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:683]: The current errors seem okayish

**Comment by magnus-madsen** [comment]
!!!jar

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:102]: Added to followup ticket

**Comment by mlutze** [comment]
@magnus-madsen What's the plan for the tests that @jaschdoc is adding? Do I merge them into my PR or do we merge them separately (second seems cleaner)

**Comment by magnus-madsen** [comment]
@mlutze All bugs allegedly fixed? Performance still OK?

**Comment by magnus-madsen** [comment]
What would you like-- that I do one more pass and one more iteration, and then we hit the big button?

**Comment by magnus-madsen** [comment]
> @magnus-madsen What's the plan for the tests that @jaschdoc is adding? Do I merge them into my PR or do we merge them separately (second seems cleaner)

@jaschdoc 

Lets merge them separately.

**Comment by magnus-madsen** [comment]
@mlutze Do we know why the flix-either package fails?

**Comment by magnus-madsen** [comment]
!!!jar

**Comment by mlutze** [comment]
Opened a PR for the community build.

**Comment by mlutze** [comment]
Performance is equivalent to master.

File: main/src/ca/uwaterloo/flix/language/ast/Scheme.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Scheme.scala:133]: Easy to add Provenance.Scheme (?).

Also could checkLessThanEqual not take a loc, so we avoid unknown which is always the absolute worst.

File: main/src/ca/uwaterloo/flix/language/ast/Scheme.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Scheme.scala:168]: Ditto about provenance and loc.

File: main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala:212]: Instead of unknown, just grab constrs.head.loc?

**Comment by magnus-madsen** [comment]
I think its pretty much ready. I am just haunting you with some SourceLocation.Unknown, but I think it is important because once we merge and if we get error reports, we really dont want unknown locations.

**Comment by mlutze** [comment]
!!!rerun

File: main/src/ca/uwaterloo/flix/language/ast/Scheme.scala
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/Scheme.scala:133]: The locations used here are essentially irrelevant, because are only put into types.

After merge, this function can change to return a Boolean rather than a result, because that's all we use in the new Typer. That means provenance is irrelevant as well.

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/Scheme.scala:133]: Added to followup

**Comment by magnus-madsen** [comment]
RFM?

**Comment by mlutze** [comment]
> RFM?

Ready!

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳 🥳

**Pull request merged**
